File: packages/sdk-js/src/breadcrumbs/buffer.ts
import type { TelemetryBreadcrumb } from '../types';

/**
 * Ring buffer for breadcrumbs
 * Automatically removes oldest entries when capacity is reached
 */
export class BreadcrumbBuffer {
  private buffer: TelemetryBreadcrumb[];
  private maxSize: number;

  constructor(maxSize: number = 100) {
    this.maxSize = Math.max(1, maxSize);
    this.buffer = [];
  }

  /**
   * Add a breadcrumb to the buffer
   */
  add(breadcrumb: Omit<TelemetryBreadcrumb, 'timestamp'>): void {
    const entry: TelemetryBreadcrumb = {
      ...breadcrumb,
      timestamp: new Date().toISOString(),
    };

    this.buffer.push(entry);

    // Remove oldest if over capacity
    while (this.buffer.length > this.maxSize) {
      this.buffer.shift();
    }
  }

  /**
   * Get all breadcrumbs (oldest first)
   */
  getAll(): TelemetryBreadcrumb[] {
    return [...this.buffer];
  }

  /**
   * Get the last N breadcrumbs
   */
  getLast(n: number): TelemetryBreadcrumb[] {
    return this.buffer.slice(-n);
  }

  /**
   * Clear all breadcrumbs
   */
  clear(): void {
    this.buffer = [];
  }

  /**
   * Get current count
   */
  get count(): number {
    return this.buffer.length;
  }

  /**
   * Get max capacity
   */
  get capacity(): number {
    return this.maxSize;
  }
}

/**
 * Helper to create a breadcrumb
 */
export function createBreadcrumb(
  type: TelemetryBreadcrumb['type'],
  category: string,
  message?: string,
  data?: Record<string, unknown>,
  level: TelemetryBreadcrumb['level'] = 'info'
): Omit<TelemetryBreadcrumb, 'timestamp'> {
  return {
    type,
    category,
    message,
    data,
    level,
  };
}

/**
 * Create an HTTP breadcrumb
 */
export function createHttpBreadcrumb(
  method: string,
  url: string,
  statusCode?: number,
  duration?: number
): Omit<TelemetryBreadcrumb, 'timestamp'> {
  const level: TelemetryBreadcrumb['level'] =
    statusCode && statusCode >= 400 ? 'error' : 'info';

  return createBreadcrumb(
    'http',
    'http',
    `${method} ${url}`,
    {
      method,
      url,
      status_code: statusCode,
      duration_ms: duration,
    },
    level
  );
}

/**
 * Create a console breadcrumb
 */
export function createConsoleBreadcrumb(
  consoleLevel: 'log' | 'info' | 'warn' | 'error' | 'debug',
  args: unknown[]
): Omit<TelemetryBreadcrumb, 'timestamp'> {
  const levelMap: Record<string, TelemetryBreadcrumb['level']> = {
    log: 'info',
    info: 'info',
    warn: 'warning',
    error: 'error',
    debug: 'debug',
  };

  return createBreadcrumb(
    'console',
    'console',
    formatConsoleArgs(args),
    { arguments: args },
    levelMap[consoleLevel] ?? 'info'
  );
}

/**
 * Create a navigation breadcrumb
 */
export function createNavigationBreadcrumb(
  from: string,
  to: string
): Omit<TelemetryBreadcrumb, 'timestamp'> {
  return createBreadcrumb(
    'navigation',
    'navigation',
    `Navigating to ${to}`,
    { from, to },
    'info'
  );
}

/**
 * Create a UI interaction breadcrumb
 */
export function createUIBreadcrumb(
  action: 'click' | 'input' | 'submit',
  target: string
): Omit<TelemetryBreadcrumb, 'timestamp'> {
  return createBreadcrumb(
    'ui',
    `ui.${action}`,
    `${action} on ${target}`,
    { target },
    'info'
  );
}

/**
 * Format console arguments to a string message
 */
function formatConsoleArgs(args: unknown[]): string {
  return args
    .map((arg) => {
      if (typeof arg === 'string') return arg;
      if (typeof arg === 'number' || typeof arg === 'boolean') return String(arg);
      if (arg === null) return 'null';
      if (arg === undefined) return 'undefined';
      try {
        return JSON.stringify(arg);
      } catch {
        return '[Object]';
      }
    })
    .join(' ');
}

File: packages/sdk-python/syntra/breadcrumbs/buffer.py
"""Ring buffer for breadcrumbs."""

from datetime import datetime, timezone
from typing import Any

from syntra.types import Breadcrumb, BreadcrumbLevel, BreadcrumbType


class BreadcrumbBuffer:
    """
    Ring buffer for breadcrumbs.

    Automatically removes oldest entries when capacity is reached.
    """

    def __init__(self, max_size: int = 100) -> None:
        self._max_size = max(1, max_size)
        self._buffer: list[Breadcrumb] = []

    def add(
        self,
        type: BreadcrumbType = BreadcrumbType.DEFAULT,
        category: str = "default",
        message: str | None = None,
        data: dict[str, Any] | None = None,
        level: BreadcrumbLevel = BreadcrumbLevel.INFO,
    ) -> None:
        """Add a breadcrumb to the buffer."""
        breadcrumb = Breadcrumb(
            type=type,
            category=category,
            message=message,
            data=data,
            level=level,
            timestamp=datetime.now(timezone.utc).isoformat() + "Z",
        )
        self._buffer.append(breadcrumb)

        # Remove oldest if over capacity
        while len(self._buffer) > self._max_size:
            self._buffer.pop(0)

    def get_all(self) -> list[Breadcrumb]:
        """Get all breadcrumbs (oldest first)."""
        return list(self._buffer)

    def get_last(self, n: int) -> list[Breadcrumb]:
        """Get the last N breadcrumbs."""
        return self._buffer[-n:]

    def clear(self) -> None:
        """Clear all breadcrumbs."""
        self._buffer = []

    @property
    def count(self) -> int:
        """Get current count."""
        return len(self._buffer)

    @property
    def capacity(self) -> int:
        """Get max capacity."""
        return self._max_size


def create_http_breadcrumb(
    method: str,
    url: str,
    status_code: int | None = None,
    duration_ms: float | None = None,
) -> Breadcrumb:
    """Create an HTTP breadcrumb."""
    level = BreadcrumbLevel.ERROR if status_code and status_code >= 400 else BreadcrumbLevel.INFO

    data: dict[str, Any] = {
        "method": method,
        "url": url,
    }
    if status_code is not None:
        data["status_code"] = status_code
    if duration_ms is not None:
        data["duration_ms"] = duration_ms

    return Breadcrumb(
        type=BreadcrumbType.HTTP,
        category="http",
        message=f"{method} {url}",
        data=data,
        level=level,
    )


def create_query_breadcrumb(
    query: str,
    duration_ms: float | None = None,
    rows_affected: int | None = None,
) -> Breadcrumb:
    """Create a database query breadcrumb."""
    data: dict[str, Any] = {"query": query}
    if duration_ms is not None:
        data["duration_ms"] = duration_ms
    if rows_affected is not None:
        data["rows_affected"] = rows_affected

    return Breadcrumb(
        type=BreadcrumbType.QUERY,
        category="db.query",
        message=query[:100] + "..." if len(query) > 100 else query,
        data=data,
        level=BreadcrumbLevel.INFO,
    )
